Implements pause audio capture

shubham1g5 opened the pull request:
## Summary

Adds a pause button in the Audio capture widget.  Change is driven by these [requirements](https://docs.google.com/document/d/11S0O_PGa90an-7s3fH953eM-OBp1XwvkxuyEpZ6n0AA/edit)


## Product Description

- Adds a pause recording feature in the Audio capture question. To activate app builder should specify the appearance as 'long' in the audio capture question. This only works on Androd 7+ devices. 

- Introduces a playback progress bar for the audio playback

- Fixes a bug in audio playback timer where after pausing and resuming the playback the timer won't show the right time elapsed for the playback. 

[See video](https://drive.google.com/file/d/1Ec8W8X5PicRqBykjHlMHrqhnJca9IfTQ/view?usp=sharing) for an overview of all the changes

## Safety Assurance

- [x] If the PR is high risk, "High Risk" label is set
- [x] I have confidence that this PR will not introduce a regression for the reasons below
- [x] Do we need to enhance manual QA test coverage ? If yes, "QA Note" label is set correctly


### Safety story

Safe changes overall. Local tested with plans to add manal QA tests. Changes are limited to the audio capture question so the radius of regressions is limited to audio capture question.

dimagimon commented:
<!--
  0 Errors
  1 Warning: This PR does not contain any J...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR does not contain any JIRA issue keys in the PR title or commit messages (e.g. KEY-123)</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dimagimon commented on app/res/layout/recording_fragment.xml:
```diff
@@ -83,20 +83,35 @@
         android:layout_weight=".1">
 
+
+        <Button
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/3ce92dc2ce8196284a31c400a04f0f4d85f2f52e/app/res/layout/recording_fragment.xml#L87"></span>Buttons in button bars should be borderless; use <code>style="?android:attr/buttonBarButtonStyle"</code> (and <code>?android:attr/buttonBarStyle</code> on the parent)</td>
    </tr>
  </tbody>
</table>

ShivamPokhriyal commented:
Thanks for sharing the video @shubham1g5. Looks great! 
I saw that the user has to save/delete recording when they try to advance to next question but I was curious, whether we can extend this and let user pause the recording and also fill out the next questions? Perhaps that'll require a much bigger change or maybe introduction of a completely new widget Or maybe that's not required by anyone as of now.

shubham1g5 commented:
@ShivamPokhriyal yeah that would be a cool feature though a bigger change to get right. We are not looking into that at the moment and the idea was just to give ability to pause the recording in case of real life interruptions.

ShivamPokhriyal commented:
Gotcha. Thanks! 
Just one additional question, what happens when user pause the recording, then puts the app in background and then come back again and resume? The reason why I'm asking this is because starting Android 9, we can't access microphone in background. And although paused state means we're not using mic, but still want to confirm the behavior.

shubham1g5 commented:
@ShivamPokhriyal Great question. That works just fine, even though the recording is on when I move out of CC. I think the background permission here means recording from some background service instead.

ctsims commented:
Hi Shubham, I have some UI feedback before jumping in on code review.

At a high level, I think the state machine for the current workflow is a bit confusing from the outside, and I'd like to avoid the irrevocable "Confirm Delete" modal if possible.

I'd like to get your high level review on how much additional work the following would be, and if it's significantly more effort, I'd still like to iterate a bit to incorporate some of the underlying state management  / verbiage from this user workflow into the existing UI, which I think should be possible even if we can't commit to the overall ui

https://confluence.dimagi.com/display/GTD/Mobile+Spec+Holders

shubham1g5 commented:
@ctsims Thanks a bunch for the design review. I will take a pass on these.

shubham1g5 commented:
@damagatchi retest this please

shubham1g5 commented:
@damagatchi retest this please

dimagimon commented on app/res/layout/audio_prototype.xml:
```diff
@@ -1,67 +1,86 @@
-            android:elevation="10dp"
-            android:layout_marginEnd="10dp" />
+            android:background="@drawable/play_disabled"
+            android:elevation="10dp" />
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/res/layout/audio_prototype.xml#L29"></span>Attribute <code>elevation</code> is only used in API level 21 and higher (current min is 16)</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/res/layout/audio_prototype.xml:
```diff
@@ -1,67 +1,86 @@
+    <LinearLayout
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="right"
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/res/layout/audio_prototype.xml#L41"></span>Use "<code>end</code>" instead of "<code>right</code>" to ensure correct behavior in right-to-left locales</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/res/layout/audio_prototype.xml:
```diff
@@ -1,67 +1,86 @@
-            android:elevation="10dp"
-            android:layout_marginEnd="30dp" />
+            android:background="@drawable/record"
+            android:elevation="10dp" />
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/res/layout/audio_prototype.xml#L74"></span>Attribute <code>elevation</code> is only used in API level 21 and higher (current min is 16)</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/res/layout/audio_prototype.xml:
```diff
@@ -1,67 +1,86 @@
             android:background="@drawable/open_file"
-            android:id="@+id/choose_file"
-            android:elevation="10dp"/>
+            android:elevation="10dp" />
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/res/layout/audio_prototype.xml#L81"></span>Attribute <code>elevation</code> is only used in API level 21 and higher (current min is 16)</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/src/org/commcare/views/widgets/CommCareAudioWidget.java:
```diff
@@ -105,88 +114,180 @@ public IAnswerData getAnswer() {
+        int seconds = (int)(((millis % (1000 * 60 * 60)) % (1000 * 60)) / 1000);
+
+        if (hours > 0) {
+            buf.append(String.format("%02d", hours)).append(":");
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L194"></span>Implicitly using the default locale is a common source of bugs: Use <code>String.format(Locale, ...)</code> instead</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/src/org/commcare/views/widgets/CommCareAudioWidget.java:
```diff
@@ -105,88 +114,180 @@ public IAnswerData getAnswer() {
+            buf.append(String.format("%02d", hours)).append(":");
+        }
+
+        buf.append(String.format("%02d", minutes)).append(":")
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L197"></span>Implicitly using the default locale is a common source of bugs: Use <code>String.format(Locale, ...)</code> instead</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/src/org/commcare/views/widgets/CommCareAudioWidget.java:
```diff
@@ -105,88 +114,180 @@ public IAnswerData getAnswer() {
+        }
+
+        buf.append(String.format("%02d", minutes)).append(":")
+                .append(String.format("%02d", seconds));
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L198"></span>Implicitly using the default locale is a common source of bugs: Use <code>String.format(Locale, ...)</code> instead</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/src/org/commcare/views/widgets/CommCareAudioWidget.java:
```diff
@@ -105,88 +114,180 @@ public IAnswerData getAnswer() {
+        player.setOnCompletionListener(mp -> onCompletePlayback());
+
+        playbackDuration.setVisibility(VISIBLE);
+        playbackDuration.setText("/" + getTimeString(player.getDuration()));
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L251"></span>Do not concatenate text displayed with <code>setText</code>. Use resource string with placeholders.</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/src/org/commcare/views/widgets/CommCareAudioWidget.java:
```diff
@@ -105,88 +114,180 @@ public IAnswerData getAnswer() {
+        playbackDuration.setText("/" + getTimeString(player.getDuration()));
+
+        playbackTime.setVisibility(VISIBLE);
+        playbackTime.setText("00:00");
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L254"></span>String literal in <code>setText</code> can not be translated. Use Android resources instead.</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/src/org/commcare/views/widgets/CommCareAudioWidget.java:
```diff
@@ -105,88 +114,180 @@ public IAnswerData getAnswer() {
+    private void onCompletePlayback() {
+        playbackSeekBar.setProgress(100);
+        stopPlaybackTimer();
+        playbackTime.setText("00:00");
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L282"></span>String literal in <code>setText</code> can not be translated. Use Android resources instead.</td>
    </tr>
  </tbody>
</table>

dimagimon commented on app/src/org/commcare/views/widgets/CommCareAudioWidget.java:
```diff
@@ -105,88 +115,180 @@ public IAnswerData getAnswer() {
+        player.setOnCompletionListener(mp -> onCompletePlayback());
+
+        playbackDuration.setVisibility(VISIBLE);
+        playbackDuration.setText("/" + getTimeString(player.getDuration()));
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/0853883e2914ad929d254cf69030a86e1bd7a364/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L252"></span>Do not concatenate text displayed with <code>setText</code>. Use resource string with placeholders.</td>
    </tr>
  </tbody>
</table>

shubham1g5 commented:
@ctsims Updated the PR description with the new design video. There is one major differences in comparison to the wireframes you designed that is there is no button to add to recording once it's been saved from the widget base screen. This feature would have required going out of Android APIs to merge the multiple files so I have skipped it for now. 

Two other pointers I am looking for feedback are - 

1. Currently pause gets enabled by setting an appearance attribute. Though there is another appearance attribute called 'acquire' that this widget supports to add a file chooser button. I can see both of them getting used together potentially though today mobile don't have support for settting multiple appearances. We can either move out the pause to the HQ app settings or support multiple appearances with a tokenised list for the audio widget. 

2. I am showing the playback progress bar irrespective of the `long` appearance. Do you see any concerns with that.

ctsims commented:
Hi Shubham,

The loss of the "Add to my recording" functionality is ok, from my perspective, because the standard workflow requires 2 user inputs (pause recording and then save recording) to 'lock in' the audio. 

A couple of points of feedback on the current flow based on the adjustments you mentioned. 

1. Once recording has started in a "Clean" state (IE: There's nothing already recorded), I think it's important that the "X" or cancel option isn't available. Ideally the box should be fully modal for the user and they shouldn't be able to interact with CommCare at all until they save the current audio recording. Even after pausing, users shouldn't be allowed to "cancel" the action. Someone who's using this to record a 50 minute audio session shouldn't be at risk of pressing the 'X' accidentally and losing the recording.

2. The way you've built the second part of the interaction (IE: "Record again" to delete the current recording) looks 90% right to me, thanks for the revisions! I think that retaining the "playback" within the dialog makes the interaction pretty confusing though.

I'd replace the microphone button in the "recorded" widget state (IE: What you see between timestamps 0:12 and 0:25) with a trash can button, and when the dialog pops up make a trash can in the "big circular button" of the dialog (which has the same function as your "Record Again" button), with a button beneath (maybe "Retain Existing Recording", but probably just "Cancel") that cancels the action. Since the user can already play the recording from the previous screen, I don't think we need the popup-dialog to retain that function. This change 

When the user presses the Big trash can button within the dialog, I think the state machine you have past 0:35 onwards (after clicking "record again" is correct (although we might want to tweak the language, since "Save" makes it sound like there is something to be saved, when in reality it commits the action of having deleted the existing recording) 

These changes add a little more confusion for a user whose intent is to change an existing recording (IE: There's no Microphone button on the XForm widget, just a delete button), but it's much safer for a user whose intent requires deleting the existing recording to need to fumble around in order to learn that they need to delete the current recording in order to do so than it is for a user whose intent was to, say, record extra audio (which isn't possible) to mistakenly think that their actions are doing so without having to have made and committed an explicit action to delete what could have been an hour long effort.

To answer your two remaining questions:

1. Totally fine to not allow uploading audio files with the "long" widget and no need to account for the two together.
2. I think this tradeoff is appropriate to reduce the code complexity. The playback bar probably isn't a value-add until about the 30-60 second mark, but these widgets are fairly special-case and if it will make it simpler to maintain them I don't think we should be putting in tons of work over-optimize about efficiency (v. safety)

shubham1g5 commented:
Thanks @ctsims That resonates with me. Here is the [new design](https://drive.google.com/file/d/1Rc1y5UaTZWLXuvnJWyrQWJm0UUFhEmco/view?usp=sharing).

dimagimon commented on app/src/org/commcare/views/widgets/CommCareAudioWidget.java:
```diff
@@ -105,88 +114,182 @@ public IAnswerData getAnswer() {
+        player.setOnCompletionListener(mp -> onCompletePlayback());
+
+        playbackDuration.setVisibility(VISIBLE);
+        playbackDuration.setText("/" + getTimeString(player.getDuration()));
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/bfbefdd1aa652ec0089d69f4acb4ae08d3543225/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L253"></span>Do not concatenate text displayed with <code>setText</code>. Use resource string with placeholders.</td>
    </tr>
  </tbody>
</table>

ctsims replied:
Very minor feedback: If this was a String.format() input, it would be easier to update / localize rather than a concatenated string

ctsims reviewed (approved):
Hey @shubham1g5 This looks great, thanks for working through all of that feedback.

A few final minor points for follow-up to close things out

* The Delete button should either be a red button with a white trashcan inside or a blue button with a white trashcan. We don't generally mix colors on icons in general, but it's especially harmful to mix Blue and Red because of a phenomenon where the cones in our eyes overlap on the spectrum. 
* I'm wondering if we should actually consider removing text from the dialogues. At the very least, I think the "Title" currently isn't adding value, which means it might be more likely to confuse than to help people. The "state" information "tap to start recording" does feel like it might reduce some ambiguity, but if we could do hallway testing with some users I wouldn't be surprised if people could figure it out on their own. 
* If we do keep the explanation texts, a few tweaks:
* The text in 0:29 should probably read "Tap to delete your recording," 
* At 30 it should probably say "Tap to start a new recording"

shubham1g5 commented:
@ctsims https://github.com/dimagi/commcare-android/pull/2534/commits/308eec35722a67885feb6f909203612a556e51f9 ,  I have removed the old header and moved the instructions to top of the dialog. Video [here](https://drive.google.com/file/d/1Ec8W8X5PicRqBykjHlMHrqhnJca9IfTQ/view?usp=sharing)

ctsims reviewed (approved):
🔥🔥🔥🔥 

Looks great

shubham1g5 commented:
@damagatchi retest this please

shubham1g5 commented:
@damagatchi retest this please
